ssl/t1_lib.c

```diff
@@ -1137,25 +1137,180 @@ static const uint16_t tls_default_sigalg[] = {
+     */
+    tmp->hash = NID_undef;
+    tmp->hash_idx = -1;
+    tmp->sig = NID_undef;
+    tmp->sig_idx = -1;
+    tmp->sigandhash = NID_undef;
+    tmp->curve = NID_undef;
+    tmp->enabled = 1;
+
+    /*
+     * Now check that the algorithm is actually usable for our property query
+     * string. Regardless of the result we still return success because we have
+     * successfully processed this sigalg, even though we may decide not to use
+     * it.
+     */
+    ret = 1;
+    keymgmt = EVP_KEYMGMT_fetch(ctx->libctx, tmp->name, ctx->propq);
+    if (keymgmt) {
+        /*
+         * We have successfully fetched the algorithm - however if the provider
+         * doesn't match this one then we ignore it.
+         *
+         * Note: We're cheating a little here. Technically if the same algorithm
+         * is available from more than one provider then it is undefined which
+         * implementation you will get back. Theoretically this could be
+         * different every time...we assume here that you'll always get the
+         * same one back if you repeat the exact same fetch. Is this a reasonable
+         * assumption to make (in which case perhaps we should document this
+         * behaviour)?
+         */
+        if (EVP_KEYMGMT_get0_provider(keymgmt) == provider) {
+            /* We have a match - so we will use this sigalg */
+           (*num_from_provider)++;
+           tmp = NULL;
+        }
+        EVP_KEYMGMT_free(keymgmt);
+    }
+ err:
+    if (tmp) {
+        OPENSSL_free(tmp->name);
+        memset(tmp, 0, sizeof(SIGALG_LOOKUP));
+    }
+    return ret;
+}
+
+static int discover_provider_sigalgs(OSSL_PROVIDER *provider, void *vctx)
+{
+    struct provider_sigalg_data_st *psd = vctx;
+
+    psd->provider = provider;
+    OSSL_PROVIDER_get_capabilities(provider, "TLS-SIGALG",
+                                   add_provider_sigalgs, psd);
+    /* If a provider returns 0 it means it does not support this capability.
+     * We should continue on looking at other providers so return 1 no matter
+     * what the function above returns.
+     */
+    return 1;
+}
+
 int ssl_setup_sig_algs(SSL_CTX *ctx)
 {
     size_t i;
     const SIGALG_LOOKUP *lu;
-    SIGALG_LOOKUP *cache
-        = OPENSSL_malloc(sizeof(*lu) * OSSL_NELEM(sigalg_lookup_tbl));
+    SIGALG_LOOKUP *from_provider = NULL;
+    size_t num_from_provider = 0;
+    struct provider_sigalg_data_st psd;
+    SIGALG_LOOKUP *cache = NULL;
     EVP_PKEY *tmpkey = EVP_PKEY_new();
     int ret = 0;
 
-    if (cache == NULL || tmpkey == NULL)
+    if (!tmpkey)
+        return 0;
+
+    psd.ctx = ctx;
+    psd.from_provider = &from_provider;
+    psd.num_from_provider = &num_from_provider;
+    psd.provider = NULL; /* Set in the discover() function. */
+    if (!OSSL_PROVIDER_do_all(ctx->libctx, discover_provider_sigalgs, &psd))
+        goto err;
+
+    cache = OPENSSL_zalloc(sizeof(*lu) * (OSSL_NELEM(sigalg_lookup_tbl) + num_from_provider));
+    if (!cache)
```

Comment:
`cache != NULL`

Probably there should be some error raised if `cache == NULL`?